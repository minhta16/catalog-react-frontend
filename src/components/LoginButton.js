import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { Button, Avatar, Menu, MenuItem, Grid, Typography } from '@material-ui/core';

export class LoginButton extends Component {
  state = {
    anchorEl: undefined,
  };

  handleClickAvatar = (e) => {
    this.setState({
      anchorEl: e.target,
    });
  };

  handleCloseMenu = () => {
    this.setState({
      anchorEl: undefined,
    });
  };

  render() {
    const { onClick, variant, color, currentUser } = this.props;
    const { anchorEl } = this.state;
    const menu = (
      <Menu
        id="simple-menu"
        anchorEl={anchorEl}
        keepMounted
        open={!!anchorEl}
        onClose={this.handleCloseMenu}
        getContentAnchorEl={null}
        anchorOrigin={{ vertical: 'bottom', horizontal: 'center' }}
        transformOrigin={{ vertical: 'top', horizontal: 'center' }}
      >
        <MenuItem onClick={this.handleCloseMenu}>Profile</MenuItem>
        <MenuItem onClick={this.handleCloseMenu}>Logout</MenuItem>
      </Menu>
    );
    return (
      <div>
        {Object.keys(currentUser).length ? (
          <Grid container alignItems="center">
            <Typography variant="body1">{`Welcome, ${currentUser.username}`}</Typography>
            <Avatar
              style={{ backgroundColor: '#fdb600', color: '#fff', margin: '0.5rem' }}
              onClick={this.handleClickAvatar}
            >
              {currentUser.username.charAt(0)}
            </Avatar>
            {menu}
          </Grid>
        ) : (
          <Button id="login-button" onClick={onClick} variant={variant} color={color}>
            Login
          </Button>
        )}
      </div>
    );
  }
}

LoginButton.propTypes = {
  onClick: PropTypes.func.isRequired,
  variant: PropTypes.string,
  color: PropTypes.string,
  currentUser: PropTypes.object.isRequired,
};

LoginButton.defaultProps = {
  variant: 'contained',
  color: 'primary',
};

export default LoginButton;
