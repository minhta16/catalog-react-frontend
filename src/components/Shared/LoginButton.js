import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import { Button, Avatar, Menu, MenuItem, Grid, Typography } from '@material-ui/core';
import PersonIcon from '@material-ui/icons/Person';
import ExitIcon from '@material-ui/icons/ExitToApp';

export class LoginButton extends Component {
  state = {
    anchorEl: undefined,
  };

  handleClickAvatar = (e) => {
    this.setState({
      anchorEl: e.target,
    });
  };

  handleCloseMenu = () => {
    this.setState({
      anchorEl: undefined,
    });
  };

  handleSignOut = () => {
    const { signOut } = this.props;
    signOut();
    this.handleCloseMenu();
  };

  render() {
    const { onClick, variant, color, currentUser } = this.props;
    const { anchorEl } = this.state;

    const menu = (
      <Menu
        id="simple-menu"
        anchorEl={anchorEl}
        keepMounted
        open={!!anchorEl}
        onClose={this.handleCloseMenu}
        getContentAnchorEl={null}
        anchorOrigin={{ vertical: 'bottom', horizontal: 'center' }}
        transformOrigin={{ vertical: 'top', horizontal: 'center' }}
      >
        <MenuItem onClick={this.handleCloseMenu} component={Link} exact="true" to="/profile">
          <Grid container alignItems="center">
            <PersonIcon color="primary" style={{ margin: '0 0.5rem 0 0' }} />
            <Typography variant="body1">Profile</Typography>
          </Grid>
        </MenuItem>
        <MenuItem
          id="login-button-logout"
          onClick={this.handleSignOut}
          component={Link}
          exact="true"
          to="/"
        >
          <ExitIcon color="primary" style={{ margin: '0 0.5rem 0 0' }} />
          <Typography variant="body1">Logout</Typography>
        </MenuItem>
      </Menu>
    );
    return (
      <div>
        {Object.keys(currentUser).length ? (
          <Grid container alignItems="center">
            <Typography variant="body1">{`Welcome, ${currentUser.username}`}</Typography>
            <Avatar
              style={{ backgroundColor: '#fdb600', color: '#fff', margin: '0.5rem' }}
              onClick={this.handleClickAvatar}
            >
              {currentUser.username.charAt(0)}
            </Avatar>
            {menu}
          </Grid>
        ) : (
          <Button id="login-button" onClick={onClick} variant={variant} color={color}>
            Login
          </Button>
        )}
      </div>
    );
  }
}

LoginButton.propTypes = {
  onClick: PropTypes.func.isRequired,
  variant: PropTypes.string,
  color: PropTypes.string,
  currentUser: PropTypes.object.isRequired,
  signOut: PropTypes.func.isRequired,
};

LoginButton.defaultProps = {
  variant: 'contained',
  color: 'primary',
};

export default LoginButton;
