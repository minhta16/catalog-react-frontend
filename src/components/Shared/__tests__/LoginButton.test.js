/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import LoginButton from 'components/Shared/LoginButton';
import { Avatar, Menu } from '@material-ui/core';

describe('components/LoginButton', () => {
  let wrapper;
  let props;
  // let warning;

  const update = () => {
    wrapper.update();
    wrapper.instance().forceUpdate();
  };

  const setup = () => {
    wrapper = shallow(<LoginButton {...props} />);
  };

  beforeEach(() => {
    props = {
      onClick: jest.fn(),
      signOut: jest.fn(),
      currentUser: {
        username: 'yolo',
        token: 'motconvitxoera2caicanh',
      },
    };
  });

  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should call signOut and handleClsoeMenu when Log out button is clicked', () => {
    setup();
    wrapper.instance().handleCloseMenu = jest.fn();
    update();
    const logout = wrapper.find('#login-button-logout');
    logout.simulate('click');
    expect(props.signOut).toHaveBeenCalled();
    expect(wrapper.instance().handleCloseMenu).toHaveBeenCalled();
  });

  it('should call handleClickAvatar when avatar is clicked', () => {
    setup();
    wrapper.instance().handleClickAvatar = jest.fn();
    update();
    const avatar = wrapper.find(Avatar);
    avatar.simulate('click');
    expect(wrapper.instance().handleClickAvatar).toHaveBeenCalled();
  });

  it('should call handleCloseMenu when menu is closed', () => {
    setup();
    wrapper.instance().handleCloseMenu = jest.fn();
    update();
    const avatar = wrapper.find(Menu);
    avatar.simulate('close');
    expect(wrapper.instance().handleCloseMenu).toHaveBeenCalled();
  });

  it('should update anchorEl when handleClickAvatar is called', () => {
    setup();
    wrapper.instance().handleClickAvatar({ target: 'meow' });
    expect(wrapper.state().anchorEl).toBe('meow');
  });

  it('should set anchorEl to undefined when handleCloseMenu is called', () => {
    setup();
    wrapper.instance().handleCloseMenu();
    expect(wrapper.state().anchorEl).toBe(undefined);
  });

  it('should show login button when currentUser is not present', () => {
    props.currentUser = {};
    setup();
    expect(wrapper.find('#login-button').length).toBe(1);
  });
});
