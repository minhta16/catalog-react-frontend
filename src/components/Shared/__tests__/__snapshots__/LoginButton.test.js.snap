// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/LoginButton should render correctly 1`] = `
<div>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
  >
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      Welcome, yolo
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Avatar))
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#9C27B0",
          "color": "#fff",
          "margin": "0.5rem",
        }
      }
    >
      y
    </WithStyles(ForwardRef(Avatar))>
    <WithStyles(ForwardRef(Menu))
      anchorOrigin={
        Object {
          "horizontal": "center",
          "vertical": "bottom",
        }
      }
      getContentAnchorEl={null}
      id="simple-menu"
      keepMounted={true}
      onClose={[Function]}
      open={false}
      transformOrigin={
        Object {
          "horizontal": "center",
          "vertical": "top",
        }
      }
    >
      <WithStyles(ForwardRef(MenuItem))
        component={[Function]}
        exact="true"
        onClick={[Function]}
        to="/profile"
      >
        <WithStyles(ForwardRef(Grid))
          alignItems="center"
          container={true}
        >
          <PersonIcon
            color="primary"
            style={
              Object {
                "margin": "0 0.5rem 0 0",
              }
            }
          />
          <WithStyles(ForwardRef(Typography))
            color="primary"
            variant="button"
          >
            Profile
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(MenuItem))>
      <WithStyles(ForwardRef(MenuItem))
        component={[Function]}
        exact="true"
        id="login-button-logout"
        onClick={[Function]}
        to="/"
      >
        <ExitToAppIcon
          color="error"
          style={
            Object {
              "margin": "0 0.5rem 0 0",
            }
          }
        />
        <WithStyles(ForwardRef(Typography))
          color="error"
          variant="button"
        >
          Logout
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(MenuItem))>
    </WithStyles(ForwardRef(Menu))>
  </WithStyles(ForwardRef(Grid))>
</div>
`;
